py/server/deephaven/experimental/keyed_record_adapter.py

```diff
@@ -0,0 +1,184 @@
+
+        ## This is a TLongIntMap:
+        row_key_to_data_key_positional_index: jpy.JType = results.rowKeyToDataKeyPositionalIndex
+
+        ## This is an int[]:
+        record_data_row_keys: typing.Sequence[int] = results.recordDataRowKeys
+
+        ## This is an Object[] (where each element is an array, of either primitives or references):
+        record_data_arrs: typing.Sequence[jpy.JType] = results.recordDataArrs
+
+        if self.is_single_key_col:
+            result_records_dict: typing.Dict[K, T] = dict()
+        else:
+            result_records_dict: typing.Dict[int, T] = dict()
+
+        # Index of current row (i.e. index into arrays of data retrieved from each DB column)
+        row_idx = 0
+        for row_key in record_data_row_keys:
+            # Find the data key corresponding to the DB row key from
+            # which this record was retrieved. Map the data key to the record.
+            data_key_idx: int = row_key_to_data_key_positional_index.get(row_key)
+            data_key_for_row = data_keys[data_key_idx]
+
+            # print(f'Building record for row_idx {row_idx}, row_key {row_key}, data_key {data_key_idx}/{data_key_for_row}')
+
+            row_data = list()
+            if self.is_single_key_col:
+                row_data.append(data_key_for_row)
+            else:
+                row_data.extend(data_key_for_row)
+
+            for col_arr in record_data_arrs:
+                # print(f'Adding data: {col_arr[row_idx]}')
+                row_data.append(col_arr[row_idx])
+
+            print(f'row_data for row {row_idx} is: {row_data}')
+            record = self.record_adapter_function(*row_data)
+
+            if self.is_single_key_col:
+                # Add the record to the results dictionary
+                result_records_dict[data_key_for_row] = record
+            else:
+                result_records_dict[data_key_idx] = record
+
+            row_idx += 1
+
+        if self.is_single_key_col:
+            return result_records_dict
+        else:
+            # Build a list with the results at the same indices as the input data keys
+            # (or None if no result was found for the requested key).
+            # This is because Python does not allow lists to be used as dictionary keys.
+            results_list: typing.List[T] = list()
+            for idx in range(0, len(data_keys)):
+                if idx in result_records_dict:
+                    results_list.append(result_records_dict[idx])
+                else:
+                    results_list.append(None)
+
+            return results_list
+
+    def _convert_keys_seq_for_java(self, data_keys: typing.Sequence, key_dtype: DType) -> typing.Sequence:
+        if type(data_keys) is numpy.ndarray:
+            # Some types (e.g. bool/DateTime) have special handling for numpy:
+            return dhnp._np_to_dh_seq(key_dtype, data_keys)
+        else:
+            return data_keys
```

Comment:
Can't `dhnp._np_to_dh_seq(` just be used in place of this whole function, since it returns the original value if it doesn't need to convert?